### PR TITLE
Add typing to `LightningModule.trainer`

### LINKED ISSUE #12337: Add typing to the Trainer object within LightningModules
## Proposed refactor

I want to add typing for the `trainer` pointer within the `LightningModule`. 

### Motivation

This would improve autocompletion for IDEs.

### Pitch
I would like to make the following addition:

(`LightningModule`)[https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/core/lightning.py#L97]
```python
class LightningModule(...):
    ...
    def __init__(self, *args: Any, **kwargs: Any) -> None:
        ...
        self.trainer = None
```

changed to 
```python
class LightningModule(...):
    ...
    def __init__(self, *args: Any, **kwargs: Any) -> None:
        ...
        self.trainer: Optional[Trainer] = None
```

### Additional context

See changes [here](https://github.com/jlhbaseball15/pytorch-lightning/commit/ee3b2428e90e8c194d63da3813413d9c8147cc04)

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @borda

### PR DESCRIPTION
**Pull request by jlhbaseball15**
## What does this PR do?

This is a small change that adds typing to the trainer object within the LightningModule.

Fixes #12337 

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -94,7 +95,7 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
         torch._C._log_api_usage_once(f"lightning.module.{self.__class__.__name__}")
 
         # pointer to the trainer object
-        self.trainer = None
+        self.trainer: Optional[Trainer] = None
```
Comment (rohitgr7) [inline pytorch_lightning/core/lightning.py:98]:
```suggestion
        self.trainer: Optional["pl.Trainer"] = None
```
I'd rather do and remove the import

**Review by awaelchli — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Review by carmocca — CHANGES_REQUESTED** [review]
Please ping me when the mypy job is green!

Thank you for your work.

File: pytorch_lightning/core/lightning.py [hunk 2]
```diff
@@ -94,7 +94,7 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
         torch._C._log_api_usage_once(f"lightning.module.{self.__class__.__name__}")
 
         # pointer to the trainer object
-        self.trainer = None
+        self.trainer: Optional["pl.Trainer"] = None
```
[the remark this answers is not in the document]
Comment (jlhbaseball15) [reply pytorch_lightning/core/lightning.py:97]: Done!

**Comment by jlhbaseball15** [comment]
@carmocca Mypy is fixed!

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/overrides/distributed.py [hunk 3]
```diff
@@ -163,5 +164,5 @@ def batch_size(self) -> int:
         return self._sampler.batch_size
 
     @property
-    def sampler(self) -> Sampler:
+    def sampler(self) -> Union[Sampler, Iterable]:
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/overrides/distributed.py:167]: It's the type defined in torch's `BatchSampler`:

`    def __init__(self, sampler: Union[Sampler[int], Iterable[int]], batch_size: int, drop_last: bool) -> None:
`
